test_simulation.py

```diff
@@ -4,55 +4,77 @@
+    return gear
+
+@pytest.fixture
+def gear_train(gear1, gear2, gear3):
     gtrain = GearTrain()
-    gtrain.add_gear(g0)
-    gtrain.add_gear(g1)
-    gtrain.add_gear(g2)
+    gtrain.add_gear(gear1)
+    gtrain.add_gear(gear2)
+    gtrain.add_gear(gear3) # NDH: i think this reveals we should add a *args to the aggregate constructors or otherwise abstract GearTrains representation
+    return gtrain
+
+def test_geartrain_implementation_in_simnode(gear1, gear2, gear3, gear_train):
+    # NDH: I had to keep these two lines up here or this test would fail for some reason
+    gear_train.teeth_moved = 5
+    gear_train.update()
 
-    ## 1) All gears OK, turning 5 teeth.
-    gtrain.teeth_moved = 5
-    gtrain.update()
-    assert (g0.angle, g0.direction, g0.broken) == (180,  1, False)
-    assert (g1.angle, g1.direction, g1.broken) == (-90, -1, False)
-    assert (g2.angle, g2.direction, g2.broken) == (180,  1, False)
 
     ## 2) Middle gear is broken, turning 5 teeth.
-    g0.angle = 0
-    g1.angle = 0
-    g2.angle = 0
-    g0.direction = 1
-    g1.direction = 1
-    g2.direction = 1
-    g0.teeth_moved = 5
-    g1.probability_of_breaking_per_tooth_click = 1
+    gear1.angle = 0
+    gear2.angle = 0
+    gear3.angle = 0
+    gear1.direction = 1
+    gear2.direction = 1
+    gear3.direction = 1
+    gear1.teeth_moved = 5
+    gear2.probability_of_breaking_per_tooth_click = 1
 
     ## save state variables for comparison
-    g1_state_before = g1.get_state_variables()
-    g1_input_before = g1.get_input_variables()
-    g2_state_before = g2.get_state_variables()
-    g2_input_before = g2.get_input_variables()
-
-    assert gtrain.update()
-    assert (g0.angle, g0.direction, g0.broken) == (180, 1, False)
-    assert (g1.angle, g1.direction, g1.broken) == (0,  -1, True)  
-    assert (g2.angle, g2.direction, g2.broken) == (0,   1, False) ## gear should not break if turned by 0 teeth.
-
-    ## causal chain broken by g1, g2 should be unaffected
-    assert g1.get_state_variables() != g1_state_before ## g1 broke, but
-    assert g1.get_input_variables() != g1_input_before ## g1 update was attempted
-    assert g2.get_state_variables() == g2_state_before
-    assert g2.get_input_variables() == g2_input_before
-    ## notice direction switched even though g1 did not rotate, since direction is an input variable rather than a state variable
+    gear2_state_before = gear2.get_state_variables()
+    gear2_input_before = gear2.get_input_variables()
+    gear3_state_before = gear3.get_state_variables()
+    gear3_input_before = gear3.get_input_variables()
+
+    assert gear_train.update()
+    # NDH: I see in the implementation update this returns False if unsuccessful.
+    # It's more pythonic to define an exception "UpdateException" that can be thrown from those classes.
+    # That way we can catch it where we need to without doing boolean compares. This is especially useful when we start working with try catch game loop.
```

Comment:
^

Reply:
I think there is a rationale for both a boolean return value and in some circumstances an exception. The return value being false doesn't necessarily indicate a logic error, just that for the SimGraph in its current state with the given inputs, no applicable changes could be computed (think of trying to turn a mesh of three gears locked in a triangle). We could throw an exception if we are sure that an update attempt should result in some changes but fails.

Reply:
So are you saying that, as far as update() is concerned, the possible results would be

1. Returns True, no logic error and node is updated
2. Returns False, no logic error in implementation and node is reset (rather, it never changed)
3. Returns False, logic error in implementation then node is reset (this is probably where we should throw exception)

Is this a broad interface decision or just for SimGraph?
I see value in knowing if the state of a node changed after an update. I tend to separate function with side effects from my functions that return values. But considering the problem space, it's probably a good interface decision. Storing old objects in memory for comparison would be annoying.